### ISSUE TITLE
split.data.table is slow? (O(n))

### ISSUE THREAD

**Issue by MichaelChirico** [Rdatatable/data.table#2950]
Was trying to create some test data for #2582 and noticed that `split.data.table` does not scale very successfully:

```
library(data.table)
library(microbenchmark)
set.seed(2034909)
sizes = sort(c(x <- 10^(1:6), 5*x))
times = sapply(sizes, function(NN) {
  DT = data.table(
    ID = rep(1:NN, sample(1:3, NN, TRUE)),
    key = 'ID'
  )
  DT[ , V1 := rnorm(.N)]
  t0 = get_nanotime()
  split(DT, by = 'ID')
  get_nanotime() - t0
})

log_n = log(sizes)
log_t = log(times)
png('~/Desktop/split_timing.png')
par(mar = c(5.1, 8.1, 4.1, 2.1))
plot(log(sizes), log(times), 
     xaxt = 'n', yaxt = 'n',
     xlab = 'Resulting Elements', ylab = '',
     main = 'Scaling of split.data.table')
x = 10^(1:7)
axis(side = 1L, at = log(x),
     labels = prettyNum(as.integer(x), big.mark = ','))
yrng = range(ceiling(log10(times) - 9))
y = 10^(yrng[1L]:yrng[2L])
axis(side = 2L, at = log(y) + 9*log(10), las = 1L,
     labels = sprintf('%02d:%02d:%05.2f', 
                      (y %/% 60) %/% 60, (y %/% 60) %% 60, y %% 60))
mtext(side = 2L, 'Time to Split (H:M:S)', line = 6L)
reg = lm(log(times) ~ log(sizes))
abline(reg, lty = 2L)
pu = par('usr')
text(pu[1] + .2*diff(pu[1:2]),
     pu[3] + .8*diff(pu[3:4]),
     labels = bquote(log(t) == .(sprintf('%.1f', coef(reg)['(Intercept)'])) + 
                       .(sprintf('%.1f', coef(reg)['log(sizes)'])) %*% log(n)))
dev.off()
```

![split_timing](https://user-images.githubusercontent.com/7606389/41806144-f8840014-76ea-11e8-82f3-59ba0c976657.png)

Basically `time ~ 200,000 * n`, so it's linear but the slope is huge.

Are we using an inefficient algorithm?

**Comment by jangorecki** [comment]
There is no any algorithms there, it is just builder for `[` call and split is made by _group by_ process. Detecting bottleneck on `setDT` might be useful, it is something which we can try to optimize. For nested we could probably avoid calling `setDT` till we reach _leafs_. But for flatten, not sure.

**Comment by jangorecki** [comment]
@MichaelChirico could you please re-run your analysis (including profiling) using `split5` branch?

**Comment by MichaelChirico** [comment]
running now, will report tomorrow

**Comment by MichaelChirico** [comment]
whoops...

> Error: vector memory exhausted (limit reached?)
> Error during wrapup: vector memory exhausted (limit reached?)

**Comment by MichaelChirico** [comment]
re-ran with `sizes = sort(c(x <- 10^(1:5), 5*x))`:

![split_timing](https://user-images.githubusercontent.com/7606389/55124631-da609e80-5141-11e9-8c36-de79d63343ca.png)

![Screen Shot 2019-03-28 at 10 23 01 AM](https://user-images.githubusercontent.com/7606389/55125074-8656b980-5143-11e9-8e71-06b74e58a446.png)

code timing still feels slow, profile looks a little different

**Comment by MichaelChirico** [comment]
Perhaps there's a memory issue? What's the size of a column pointer for `alloc.col`? I guess `split.data.table` will be requesting `1024*length(split(DT))` column pointers' memory to be set aside...

**Comment by MichaelChirico** [comment]
I changed `split.data.table` to use `alloc.col(d, n = 32L)` & I think that sped it up (still 30 seconds feels a bit slow):

![Screen Shot 2019-03-28 at 10 41 00 AM](https://user-images.githubusercontent.com/7606389/55125874-1f86cf80-5146-11e9-854d-79a688747764.png)

This also might be an issue where `getOption` is being called for every item in the list as well... so could also set `n = getOption('datatable.alloccol')` and `verbose = getOption('datatable.verbose')` outside the loop and pass it through.

**Comment by MichaelChirico** [comment]
Version setting `n` and `verbose` outside the loop to the defaults & passing through:

![Screen Shot 2019-03-28 at 10 46 55 AM](https://user-images.githubusercontent.com/7606389/55126067-e438d080-5146-11e9-9c69-dc15ba3b32d6.png)

An improvement, but using a lower `n` for `alloc.col` is 2x as fast it seems.

I would lean towards using a lower `n` (`max(32L, 2L*ncol(d))` could be more robust).

I guess the use case of running `split` and then adding hundreds of columns is quite rare, much more common to add the columns before splitting.

**Comment by jangorecki** [comment]
I see a valid use case when for different groups of rows user want to add different numbers of columns, then split followed by `:=` is valid use case. If pre-allocating 32 cols is only twice faster than 1024 then it is not that much beneficial. Also it can be confusing for users that normally 1024 cols are pre-allocated and here only 32. It is difficult to well document such exception.
I now think the ultimate solution is to push down `split` to C and then pre-allocate from there. This is basically and inverse process of `rbindlist` which was recently reworked. Lets wait and see what is Matt's opinion on that.

**Comment by mattdowle** [comment]
What's the use-case for `split(..., flatten=FALSE)`?

**Comment by MichaelChirico** [comment]
> But why not do that using: `DT[, fwrite(.SD), by=...]`

you're right that it's better, at a glance I didn't come up with a more relevant use case. If we can't improve the scaling of `split` we can just add a caveat to the manual to consider doing things `by` group rather than adding hundreds of `data.table`s to the namespace I guess?

**Comment by jangorecki** [comment]
> What's the use-case for `split(..., flatten=FALSE)`?

it is the default one (for consistency to base split which doesn't have `flatten=FALSE`), when you split by multiple variables it will return non-nested list.
